### PR TITLE
bpo-39200: Correct the error message for range() empty constructor

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue39200

### CODE REVIEW

**Review by corona10 — APPROVED** [review]
For me, this is more correct :)

File: Objects/rangeobject.c [hunk 1]
```diff
@@ -78,8 +78,12 @@ range_new(PyTypeObject *type, PyObject *args, PyObject *kw)
         return NULL;
 
     if (PyTuple_Size(args) <= 1) {
-        if (!PyArg_UnpackTuple(args, "range", 1, 1, &stop))
+        if (!PyArg_UnpackTuple(args, "range", 1, 1, &stop)) {
```
Comment (serhiy-storchaka) [inline Objects/rangeobject.c]: I think it is better to get rid of `PyArg_UnpackTuple()`. Its only benefit is checking the size of the tuple and formatting an error message which is wrong. ~Note also that the second `PyArg_UnpackTuple()` also can produce incorrect error message.~
```c
switch (PyTuple_GET_SIZE(args)) {
case 3:
    step = PyTuple_GET_ITEM(args, 2);
    /* fall through */
case 2:
    start = PyTuple_GET_ITEM(args, 0);
    stop = PyTuple_GET_ITEM(args, 1);
    ...
    break;
case 1:
    stop = PyTuple_GET_ITEM(args, 0);
    ...
    break;
case 0:
    // error
    return NULL;
default:
    // error
    return NULL;
}
```

**Review by serhiy-storchaka — COMMENTED** [review]
Would be nice to add tests.

**Comment by serhiy-storchaka** [comment]
Other option - just use `PyArg_UnpackTuple(args, "range", 1, 3, &start, &stop, &step)` and check if `stop == NULL`.

**Comment by pablogsal** [comment]
> Other option - just use `PyArg_UnpackTuple(args, "range", 1, 3, &start, &stop, &step)` and check if `stop == NULL`.

I have implemented the switch statement as we would need to do the dance with `step` and `start` anyway and I find the switch statement a bit more elegant.

File: Objects/rangeobject.c (hunk 1 above)
Comment (pablogsal) [reply Objects/rangeobject.c]: Thanks for the suggestion, @serhiy-storchaka!

**Comment by pablogsal** [comment]
> Would be nice to add tests.

Done in 146591c

**Comment by pablogsal** [comment]
Btw, as a bonus the constructor of `range` is anecdotally faster:

# master
 ./python.exe -m pyperf timeit "range(1,3,4)"
.....................
Mean +- std dev: 893 ns +- 44 ns

# patched

 ./python.exe -m pyperf timeit "range(1,3,4)"
.....................
Mean +- std dev: 822 ns +- 14 ns

File: Lib/test/test_range.py [hunk 2]
```diff
@@ -91,6 +91,13 @@ def test_range(self):
         r = range(-sys.maxsize, sys.maxsize, 2)
         self.assertEqual(len(r), sys.maxsize)
 
+    def test_range_constructor_error_messages(self):
+        with self.assertRaises(TypeError, msg="range expected at least 1 argument, got 0"):
```
Comment (corona10) [inline Lib/test/test_range.py]: If this code is intended to check whether the error message is correct,
I'd like to suggest to use `assertRaisesRegex`.

File: Lib/test/test_range.py [hunk 3]
```diff
@@ -91,6 +91,13 @@ def test_range(self):
...
+            range()
+
+        with self.assertRaises(TypeError, msg="range expected at most 3 arguments, got 6"):
```
Comment (corona10) [inline Lib/test/test_range.py]: Same here

**Review by corona10 — COMMENTED** [review]
@pablogsal I left comments on test code :)

File: Lib/test/test_range.py (hunk 2 above)
Comment (pablogsal) [reply Lib/test/test_range.py]: Woops! Thanks for the catch!

**Review by corona10 — APPROVED** [review]

File: Objects/rangeobject.c [hunk 4]
```diff
@@ -77,37 +77,49 @@ range_new(PyTypeObject *type, PyObject *args, PyObject *kw)
...
+            Py_INCREF(_PyLong_One);
+            step = _PyLong_One;
+            break;
+        case 0:
+            PyErr_SetString(PyExc_TypeError, "range expected at least 1 argument, got 0");
             return NULL;
-        stop = PyNumber_Index(stop);
-        if (!stop)
+        default:
+            PyErr_Format(PyExc_TypeError, "range expected at most 3 arguments, got %i", num_args);
```
Comment (serhiy-storchaka) [inline Objects/rangeobject.c]:
```suggestion
            PyErr_Format(PyExc_TypeError, "range expected at most 3 arguments, got %zd", num_args);
```

**Review by serhiy-storchaka — APPROVED** [review]

File: Lib/test/test_range.py [hunk 5]
```diff
@@ -91,6 +91,13 @@ def test_range(self):
         r = range(-sys.maxsize, sys.maxsize, 2)
         self.assertEqual(len(r), sys.maxsize)
 
+    def test_range_constructor_error_messages(self):
+        with self.assertRaisesRegex(TypeError, "range expected at least 1 argument, got 0"):
```
Comment (ZackerySpytz) [inline Lib/test/test_range.py]: Why doesn't this code follow PEP 8?

File: Lib/test/test_range.py [hunk 6]
```diff
@@ -91,6 +91,13 @@ def test_range(self):
...
+            range()
+
+        with self.assertRaisesRegex(TypeError, "range expected at most 3 arguments, got 6"):
```
Comment (ZackerySpytz) [inline Lib/test/test_range.py]: Ditto.

File: Objects/rangeobject.c [hunk 7]
```diff
@@ -77,37 +77,49 @@ range_new(PyTypeObject *type, PyObject *args, PyObject *kw)
...
+            if (!stop) {
+                return NULL;
+            }
+            Py_INCREF(_PyLong_Zero);
+            start = _PyLong_Zero;
+            Py_INCREF(_PyLong_One);
+            step = _PyLong_One;
+            break;
+        case 0:
+            PyErr_SetString(PyExc_TypeError, "range expected at least 1 argument, got 0");
```
Comment (ZackerySpytz) [inline Objects/rangeobject.c]: This line is too long.

File: Objects/rangeobject.c [hunk 8]
```diff
@@ -77,37 +77,49 @@ range_new(PyTypeObject *type, PyObject *args, PyObject *kw)
...
             return NULL;
-        stop = PyNumber_Index(stop);
-        if (!stop)
+        default:
+            PyErr_Format(PyExc_TypeError, "range expected at most 3 arguments, got %zd", num_args);
```
Comment (ZackerySpytz) [inline Objects/rangeobject.c]: This line is too long.

File: Objects/rangeobject.c [hunk 9]
```diff
@@ -77,37 +77,49 @@ range_new(PyTypeObject *type, PyObject *args, PyObject *kw)
     if (!_PyArg_NoKeywords("range", kw))
         return NULL;
 
-    if (PyTuple_Size(args) <= 1) {
-        if (!PyArg_UnpackTuple(args, "range", 1, 1, &stop))
+    Py_ssize_t num_args = PyTuple_GET_SIZE(args);
+    switch (num_args) {
+        case 3:
+            step = PyTuple_GET_ITEM(args, 2);
+            /* FALLTHRU */
```
Comment (ZackerySpytz) [inline Objects/rangeobject.c]: I think it would be better to use the spelling "fallthrough".

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**